pkg/client/connector/install.go

```diff
@@ -372,202 +403,290 @@ func findMatchingPort(dep *kates.Deployment, portName string, svcs []*kates.Serv
+// is installed alongside the proper workload. In doing that, it also ensures that
+// the workload is referenced by a service. Lastly, it returns the service UID
+// associated with the workload since this is where that correlation is made.
+func (ki *installer) ensureAgent(c context.Context, namespace, name, svcName, portName, agentImageName string) (string, string, error) {
+	kind, err := ki.findObjectKind(c, namespace, name)
 	if err != nil {
-		if errors2.IsNotFound(err) {
-			err = agentNotFound
+		return "", "", err
+	}
+	var obj kates.Object
+	switch kind {
+	case "ReplicaSet":
+		obj, err = ki.findReplicaSet(c, namespace, name)
+		if err != nil {
+			return "", "", err
 		}
-		return "", err
+	case "Deployment":
+		obj, err = ki.findDeployment(c, namespace, name)
+		if err != nil {
+			return "", "", err
+		}
+	default:
+		return "", "", fmt.Errorf("Cannot ensure agent on unsupported workload: %s", kind)
+	}
+
+	podTemplate, _, err := GetPodTemplateFromObject(obj)
+	if err != nil {
+		return "", "", err
 	}
 	var agentContainer *kates.Container
-	for i := range dep.Spec.Template.Spec.Containers {
-		container := &dep.Spec.Template.Spec.Containers[i]
+	for i := range podTemplate.Spec.Containers {
+		container := &podTemplate.Spec.Containers[i]
 		if container.Name == agentContainerName {
 			agentContainer = container
 			break
 		}
 	}
 
-	svcPortChanged, err := portChanged(c, dep, portName)
-	if err != nil {
-		return "", err
+	if err := checkSvcSame(c, obj, svcName, portName); err != nil {
+		msg := fmt.Sprintf(
+			`%s already being used for intercept with a different service
+			configuration. To intercept this with your new configuration,
+			please use telepresence uninstall --agent %s This will cancel 
+			any intercepts that already exist for this service`, kind, obj.GetName())
+		return "", "", errors.Wrap(err, msg)
 	}
 	var svc *kates.Service
 
 	switch {
-	case svcPortChanged:
-		dlog.Infof(c, "Port to be intercepted changed for deployment %s.%s", name, namespace)
-		dlog.Info(c, "Undoing dep+svc modifications and removing agent")
-		// Remove deployment+svc modifications, as well as traffic-agent since
-		// the port is vital in configuring all of those
-		if err = ki.undoDeploymentMods(c, dep); err != nil {
-			return "", err
-		}
-
-		if err = ki.waitForApply(c, namespace, name, dep); err != nil {
-			return "", err
-		}
-		// Since the agent has been removed, we find the updated version of
-		// the deployment and fallthrough to the no agent case
-		dep, err = ki.findDeployment(c, namespace, name)
-		if err != nil {
-			return "", err
-		}
-		fallthrough
 	case agentContainer == nil:
 		dlog.Infof(c, "no agent found for deployment %s.%s", name, namespace)
 		dlog.Infof(c, "Using port name %q", portName)
-		matchingSvcs := ki.findMatchingServices(portName, dep)
+		matchingSvcs := ki.findMatchingServices(portName, svcName, podTemplate.Labels)
 		if len(matchingSvcs) == 0 {
```

Comment:
Does it make sense here to, similar to when a service has more than one port we tell the user they have to specify a port, to tell users if they have more than one service matching their workload, they also need to specify the service?  I'm not sure why we ever wanted to proceed when we match more than one service to begin with.

Reply:
Yes, I think it sounds reasonable to report ambiguities. Especially if the user can remove them by providing more info.